ntripping-cli/src/main.rs

```diff
@@ -0,0 +1,183 @@
+    #[clap(long, default_value_t = 0)]
+    request_counter: u8,
+
+    /// Area ID to be used in generation of CRA message. If this flag is set, ntripping outputs messages of type CRA rather than the default GGA
+    #[clap(long)]
+    area_id: Option<u32>,
+
+    /// Field specifying which types of corrections are to be received
+    #[clap(long)]
+    corrections_mask: Option<u16>,
+
+    /// Solution ID, the identifier of the connection stream to reconnect to in the event of disconnections
+    #[clap(long)]
+    solution_id: Option<u8>,
+}
+
+type Result<T> = std::result::Result<T, Box<dyn std::error::Error + Send + Sync>>;
+
+async fn run(opt: Cli) -> Result<()> {
+    let now = || -> DateTime<Utc> {
+        if let Some(epoch) = opt.epoch {
+            SystemTime::UNIX_EPOCH + Duration::from_secs(epoch.into())
+        } else {
+            SystemTime::now()
+        }
+        .into()
+    };
+
+    let auth = if let (Some(username), Some(password)) = (opt.username, opt.password) {
+        Some(Auth::new(username, password))
+    } else {
+        None
+    };
+
+    let msg: Sentence = if let Some(area_id) = opt.area_id {
+        Cra::new()
+            .with_area_id(area_id)
+            .with_corrections_mask(opt.corrections_mask)
+            .with_solution_id(opt.solution_id)
+            .into()
+    } else {
+        Gga::new()
+            .with_time(now())
+            .with_lat(opt.lat)
+            .with_lon(opt.lon)
+            .with_height(opt.height)
+            .into()
+    };
+
+    let client = {
+        let client = Client::new().with_client_id(opt.client_id).with_auth(auth);
+        if opt.gga_header {
+            client.with_ntrip_gga(msg)
+        } else {
+            client
+        }
+    };
+
+    let (mut sink, mut stream) = {
+        let url = if !opt.url.starts_with("http://") && !opt.url.starts_with("https://") {
+            format!("https://{}", opt.url)
+        } else {
+            opt.url
+        };
+        let uri = url.parse()?;
+        client.connect(uri).await?.split()
+    };
+
+    let writer_task = tokio::spawn(async move {
+        let mut out = io::stdout();
+        while let Some(data) = stream.try_next().await? {
+            out.write_all(&data).await?;
+        }
+        Result::Ok(())
+    });
+
+    if opt.gga_period == 0 {
+        return writer_task.await?;
+    }
+
+    let mut request_counter = opt.request_counter;
+    let mut gga_interval = tokio::time::interval(Duration::from_secs(opt.gga_period));
+    loop {
+        if writer_task.is_finished() {
+            break;
+        }
+        gga_interval.tick().await;
+        let sentence = msg.with_time(now()).with_request_counter(request_counter);
+        let _ = sink.send(sentence).await;
+        request_counter = request_counter.wrapping_add(1);
+    }
+
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    let opt = Cli::parse();
+    tracing_subscriber::fmt::fmt()
+        .with_max_level(match opt.verbose {
```

Comment:
For this to work, I think we need to set `RUST_LOG` and possible add a logger? `env_logger`?  As it right now the `-v` option isn't showing anything for me.

Reply:
It shouldn't need it set, try `-vv`? I guess I have nothing logging at the debug level

Reply:
Actually looks like I do, and I see them with just `-v`

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/ntripping -v`
2022-12-15T22:54:51.013819Z DEBUG ntripping: connect addr=na.skylark.swiftnav.com:2101
```

Reply:
Looks like trace output is going to stdout

![image](https://user-images.githubusercontent.com/183436/207988514-6c3ebdc1-938b-44c8-8455-cb5b7db5219a.png)

Reply:
A little concerned that we're losing some user facing features, like connection logging from cURL:

![image](https://user-images.githubusercontent.com/183436/207989497-6867fcec-fc2a-499c-8bd8-fd833dc3f5ca.png)

Versus:

![image](https://user-images.githubusercontent.com/183436/207989580-93cf42ae-f522-4f73-ae49-67c2d392b4d4.png)

So we're loosing logging around a few things:
- Response headers
- GGA/CRA string sent up
- Connection state

Reply:
There is not rush on this probably should have made this pr a draft. the metrics project I wanna use this for just got pushed back again. Just have been missing rust a bit

I also still want to add
- connection pooling so requests to different mountpoints but the same host share a connection
- automatic retry, plus a way to hook into the retries (so you can, for example, have the retry send a CRA which a certain correction mask)
- maybe sourcetable types? I was thinking it might be nice to just point this too-be-built app at a caster and have it discover what mountpoints to monitor

So definitely can make it feature complete with the old version. Could also try a version with curl, but crate we use now is pretty much all `!Send + !Sync` which is a little annoying

Reply:
👍

Reply:
Maybe https://docs.rs/isahc/latest/isahc/ would work as an alternative?  It's purports to be a more "Rusty" wrapper of cURL.

Reply:
Looks like there's no easy way to enable HTTP 0.9 in this library though, so might have to be hacked in.

Reply:
Yeah that's the first library I tried but saw it had and old open issue about that, but maybe it wouldn't be too hard to add